app/service/reporting_svc.py

```diff
@@ -0,0 +1,126 @@
+        agents_steps = {a['paw']: {'agent_id': a['id'], 'steps': []} for a in op['host_group']}
+        for step in op['chain']:
+            ability = (await self.data_svc.explode_abilities(criteria=dict(id=step['ability'])))[0]
+            command = self.decode_bytes(step['command'])
+            step_report = dict(ability_id=ability['ability_id'],
+                               command=command,
+                               delegated=step['collect'],
+                               run=step['finish'],
+                               status=step['status'],
+                               platform=ability['platform'],
+                               executor=step['executor'],
+                               description=ability['description'],
+                               name=ability['name'],
+                               attack=dict(tactic=ability['tactic'],
+                                           technique_name=ability['technique_name'],
+                                           technique_id=ability['technique_id'])
+                               )
+            if agent_output:
+                result = (await self.data_svc.explode_results(criteria=dict(link_id=step['id'])))[0]
+                step_report['output'] = self.decode_bytes(result['output'])
+            agents_steps[step['paw']]['steps'].append(step_report)
+        report['steps'] = agents_steps
+        report['skipped_abilities'] = await self.get_skipped_abilities_by_agent(op_id=op['id'])
+        return report
+
+    async def get_skipped_abilities_by_agent(self, op_id):
+        """
+        Generate a list of skipped abilities for agents in an operation
+        :param op_id: operation id
+        :return: a JSON skipped abilities list by agent
+        """
+        op_facts, op_results, op_state, op_group, op_adversary = await self._get_operation_skipped_ability_data(op_id)
+        abilities_by_agent = await self._get_all_possible_abilities_by_agent(hosts=op_group, adversary=op_adversary)
+        skipped_abilities = []
+        for agent in op_group:
+            agent_skipped = defaultdict(dict)
+            agent_executors = [a['executor'] for a in agent['executors']]
+            agent_ran = set([(await self.data_svc.explode_abilities(dict(id=ab)))[0]['ability_id'] for ab in op_results[agent['paw']]])
+            for ab in abilities_by_agent[agent['paw']]['all_abilities']:
+                skipped = await self._check_reason_skipped(agent=agent, ability=ab, op_facts=op_facts, state=op_state,
+                                                           agent_executors=agent_executors, agent_ran=agent_ran)
+                if skipped:
+                    if agent_skipped[skipped['ability_id']]:
+                        if agent_skipped[skipped['ability_id']]['reason_id'] < skipped['reason_id']:
+                            del agent_skipped[skipped['ability_id']]
```

Comment:
is this del necessary or will the next line overwrite the entry?

Reply:
Unnecessary and removed